kv_storage.h

```diff
@@ -0,0 +1,383 @@
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef SHERLOCK_KV_STORAGE_H
+#define SHERLOCK_KV_STORAGE_H
+
+#include <atomic>
+#include <future>
+#include <string>
+
+#include "sherlock.h"
+
+#include "../Bricks/mq/inmemory/mq.h"
+
+namespace sherlock {
+namespace kv_storage {
+// Base structures for user entry types.
+struct AllowNonThrowingGet {
+  constexpr static bool allow_nonthrowing_get = true;
+};
+
+enum NullEntryTypeHelper { NullEntry };
+struct Nullable {
+  const bool exists;
+  Nullable() : exists(true) {}
+  Nullable(NullEntryTypeHelper) : exists(false) {}
+};
+
+struct Deletable {};  // User promises to serialize Nullable::exists;
+
+// Universal null entry creation.
+template <typename T_ENTRY, bool IS_NULLABLE>
+struct GetNullEntry {
+  static T_ENTRY get() {
+    static_assert(true, "Should never happen.");
+    return T_ENTRY();
+  }
+};
+
+template <typename T_ENTRY>
+struct GetNullEntry<T_ENTRY, true> {
+  static T_ENTRY get() {
+    T_ENTRY e(NullEntry);
+    return e;
+  }
+};
+
+template <typename T_ENTRY, bool IS_NULLABLE>
+T_ENTRY CreateNullEntry() { return GetNullEntry<T_ENTRY, IS_NULLABLE>::get(); }
+
+// Exceptions.
+struct KeyNotFoundException : bricks::Exception {};
+
+template <typename T_ENTRY>
+struct SpecificEntryTypeKeyNotFoundException : KeyNotFoundException {
+  typedef typename std::remove_reference<decltype(std::declval<T_ENTRY>().key())>::type T_KEY;
+  T_KEY key;
+  SpecificEntryTypeKeyNotFoundException(const T_KEY& key) : key(key) {}
+};
+
+struct KeyAlreadyExistsException : bricks::Exception {};
+
+template <typename T_ENTRY>
+struct SpecificEntryKeyAlreadyExistsException : KeyAlreadyExistsException {
+  T_ENTRY entry;
+  SpecificEntryKeyAlreadyExistsException(const T_ENTRY& entry): entry(entry) {}
+};
+
+struct EntryShouldExistException : bricks::Exception {};
+
+template <typename T_ENTRY>
+struct SpecificEntryShouldExistException : EntryShouldExistException {
+  T_ENTRY entry;
+  SpecificEntryShouldExistException(const T_ENTRY& entry): entry(entry) {}
+};
+
+// API calls, `[Async]{Get/Add/Update/Delete}(...)`.
+//
+// C++ API supports three syntaxes:
+//
+//   1) Asynchronous, using std::{promise/future}.
+//      To use the async/await paradigm.
+//      `AsyncAdd(entry)` and `AsyncGet(key)`, both return a future. The `AsyncGet()` one may throw.
```

Comment:
Missing comments on 1) conditions under which `AsyncGet()` throws and how to avoid it, and 2) conditions under which `AsyncAdd()` throws and how to avoid it.